### ISSUE TITLE
Will cilium IPAM support specified ip address of pod?

### ISSUE THREAD

**Issue by xgp01** [cilium/cilium#17026]
## Proposal / RFE

**Is your feature request related to a problem?**

**Describe the solution you'd like**

As the title states, in some cases we need a static pod ip address because our program does not support DNS.  We think if a pod can be assign with a static ip address will solve this problem. We wonder if cilium will support it in future versions or maybe there're better ways we haven't found? 
We found that calico IPAM has a feature by setting annotation "cni.projectcalico.org/ipAddrs" can specific ip address to single pod, but cilium has more other powerful features that we need, so we decide use cilium as our network plugin.

Looking forward to your favorable  reply.

**Comment by vsxen** [comment]
a static pod ip?

- your pod only a replica？
- your pod don't need rollingUpdate?

if then please use vm instead of container.

**Comment by XieGP** [comment]
> a static pod ip?
> 
> * your pod only a replica？
> * your pod don't need rollingUpdate?
> 
> if then please use vm instead of container.

Thanks for your suggestion!
Our pod is indeed only one replica, but we need kubernetes to schedule it and update images.  And other components are deployed in kubernetes, we want this system deploy in same achitechture.

**Comment by aanm** [comment]
Why can't the pod be behind a k8s service and have a static IP address in the service instead?

**Comment by jbowen93** [comment]
Bumping this because it's something I would also like. I'm not sure about XieGP's needs, but my needs relate to libp2p where I need the outbound and inbound IP address of a pod to match. I'm also currently using Calico to manage this. 

The Protocol Labs team (creators of libp2p) have a tool called [Testground](https://github.com/testground/testground) that is used for large scale (up to 50k nodes) testing of libp2p and other p2p protocols. This tool utilizes Kubernetes and their [current install instructions ](https://github.com/testground/infra/blob/master/k8s/01_install_k8s.sh) use Weave which [supports this feature](https://www.weave.works/docs/net/latest/tasks/ipam/manual-ip-address/). 

For a lot of use cases doing a pod level SNAT is probably sufficient, but for quick integration testing it's nice that I can just spin up a single node minikube cluster. For the Protocol Labs Testground use case it's easier to orchestrate 50k containers with K8s and CNI than it is with VMs.

**Comment by akamac** [comment]
My use case is to run a static pod, not in the hostNet, but be able to access it even before apiserver is up.

**Comment by pulcov** [comment]
Hello and thank you for your work.
I use this feature with calico to assign an ip to a namespace ([https://projectcalico.docs.tigera.io/networking/legacy-firewalls](https://projectcalico.docs.tigera.io/networking/legacy-firewalls)) or a pod ([https://projectcalico.docs.tigera.io/networking/use-specific-ip](https://projectcalico.docs.tigera.io/networking/use-specific-ip)).
My router is configured to do traffic shaping based on source IPs. The implementation in cilium would really be a positive point for mixed environments (kubernetes and legacy)

**Comment by m-yosefpor** [comment]
> a static pod ip?
> 
>     * your pod only a replica？
> 
>     * your pod don't need rollingUpdate?
> 
> 
> if then please use vm instead of container.

If we want to run VMs in k8s, solutions such as kubevirt are also working with CNI for networking. So this feature will help having VMs with static IPs for kubevirt.

**Comment by m-yosefpor** [comment]
> Why can't the pod be behind a k8s service and have a static IP address in the service instead?

This way only the ingress traffic will have the clusterIP, but egress traffic does not have that IP. Also some processes are working with interface IP set on interfaces (such as database clustering for advertised IP), so for such static usecases, being able to assign a specific IP to pods are useful.


I'm not sure how is this possible to be implemented. I will take a look at Calico's implementation. In VXLAN mode it's more straight forward to implement this, but in direct routing or BGP mode, pods IPs assigned to nodes have deterministic IPs from node CIDR pool, so if the same pod is scheduled to other nodes, these static routes (for direct routing) should be updated, (since pod ip is `/32` and is more specific than podCIDR for nodes (`/14` by default), then the traffic is routed to correct node, however in BGP mode, I'm not sure what is the behavior of routers. Will they route traffic to more specific route, or will they multi-path traffic to both nodes? If it's the latter case, maybe cilium should check to ensure the pod static IP is not from already assigned pod ranges to nodes, and they should be from another range.

**Comment by eburghar** [comment]
There is a common use-case for this. Sending email with ipv6 impose to have a PTR in the reverse zone. As external-dns does not handle ipv6 yet, and provider allowing the creation of PTR through API or delegating the ipv6 reverse zone (to use your own dns servers) are exceptions, we are stick to manually edit the PTR each time the IP of the MTA pod change.

Fixing IP or specifying a big cidr (/125) so we could provision PTR entries by advance would be nice to have

**Comment by xiaods** [comment]
I don't think this is feature, this is customized feature.

**Comment by oscni** [comment]
If/when this issue is fixed please consider to not only allow to announce /32 but instead make it more flexible, basically with the ability to create in IPAM a bigger network/IPPool, lets say a /24 which we can for an example call "application1". This /24 can be used for in external firewall rules or servers ip-filter where you need to know who the source is.
Then also allow to specify a subnetSize/blockSize which indicates to the nodes how large of a subnet they should dynamically allocate from the IPPool when they spin up a pod, for an example a /28 makes every node get at least 16 IPs. If one single node needs more than the IPs from 1 subnet it could of course allocate more subnets.

With this solution Egress Gateway wont be needed and instead IPs can be used directly without SNAT/DNAT/PAT. 🚀 🚀 🚀

**Comment by karibertils** [comment]
Would love to see the possibility to specify which CIDR's is used for a set of pods, or all pods in a namespace.

**Comment by gandro** [comment]
There is currently no concrete design or plan to support allocating a specific static IP to a pod. Egress gateway is currently probably the closest to what we have right now.

What we _are_ currently planning is the ability to have certain workloads allocate an IP from different subnet pool than the default pod CIDR.

**Comment by ysksuzuki** [comment]
Hi, @gandro 

> What we are currently planning is the ability to have certain workloads allocate an IP from different subnet pool than the default pod CIDR.

Would you be able to elaborate on that feature? Actually, I'm planning to propose implementing Namespace scope IPAM to cover our use cases, but the feature you mentioned might be similar to it. The namespace scope IPAM is an IPAM mode in which Cilium allocates IP addresses for pods in a certain namespace from a different address pool associated with the namespace than the default pod CIDR.

**Comment by ysksuzuki** [comment]
@gandro Thank you for sharing it!

I think the feature you are working on could cover our use cases if users can assign a particular pool to a specific namespace. Is the design document available somewhere?

**Comment by gandro** [comment]
> I think the feature you are working on could cover our use cases if users can assign a particular pool to a specific namespace. Is the design document available somewhere?

Yes, we have a design document. It's slightly outdated however, but it should cover the basic design:

https://docs.google.com/document/d/1FIyfhVegljG9anXpyA25gEYzuUGpfHo_MycZ-Wk0VeE/edit#heading=h.5mwz58y9syps

> If is there anything I could do to have the pool gets assigned to the namespace or something, I'm happy to help!

Will keep this in mind. I think around the assignment policies there is opportunity for contribution

**Comment by ysksuzuki** [comment]
Thanks! I'm looking forward to seeing this feature land.

**Comment by xagent003** [comment]
> a static pod ip?
> 
>     * your pod only a replica？
> 
>     * your pod don't need rollingUpdate?
> 
> 
> if then please use vm instead of container.

Kubevirt VMs are run in a Pod/container

**Comment by MohammedNoureldin** [comment]
I am excited to see a commit related to this! Does this mean, this feature is coming soon?

**Comment by gandro** [comment]
> I am excited to see a commit related to this! Does this mean, this feature is coming soon?

A variant of this feature has been merged as part of the multi-pool IPAM feature - one can now specify a pool of IPs and assign them to particular workloads, see the docs here: https://docs.cilium.io/en/v1.14/network/concepts/ipam/multi-pool/

It currently only supports assigning IPs out of a range of IPs, pools with only a single IPs are currently not supported. But that might change in v1.15 once we have fixed some technical shortcomings of the IPAM allocator here: https://github.com/cilium/cilium/pull/26488

**Comment by MohammedNoureldin** [comment]
Thank you, @gandro!

I do not have access to my computer for a couple of weeks to test this. Does this mean, that using x.x.x.x/32 as a pool won't work currently to specify an exact IP address for a pod?

**Comment by gandro** [comment]
>  Does this mean, that using x.x.x.x/32 as a pool won't work currently to specify an exact IP address for a pod?

Yes, that's correct. Once https://github.com/cilium/cilium/pull/26488 is merged, fixing this shortcoming should be fairly easy. But for now, it's not supported.

**Comment by lingdie** [comment]
👀 Any updates?

**Comment by gandro** [comment]
We've since identified that https://github.com/cilium/cilium/pull/26488 might not be straightforward after all. I've wrote them down here: https://github.com/cilium/cilium/issues/28637#issuecomment-1777128063

In particular, I still think we want to move over to the new allocator, but it needs support for the random allocation strategy first.

Contributions welcome!

**Comment by evenh** [comment]
At KubeCon EU 2024 [crik](https://github.com/qawolf/crik) was mentioned in [this talk](https://www.youtube.com/watch?v=c2MbSM9-7Xs) as a project that could benefit form spawning new pods with a fixed IP when restoring state from a previous pod that has been shutdown.

**Comment by muvaf** [comment]
@evenh Thanks for bringing up!

[`crik`](https://github.com/qawolf/crik) uses [`criu`](https://github.com/checkpoint-restore/criu) under the hood to trigger checkpoint and restore. I think checkpoint/restore is a good use case for pre-specificed IP for pods, see https://github.com/checkpoint-restore/criu/issues/211 from criu - IP change effectively renders the live migration impossible since it forces dropping the connections.

On another note, upstream Kubernetes introduces [support](https://kubernetes.io/blog/2023/03/10/forensic-container-analysis/) for using `criu` to take checkpoint and restore so it's mostly matter of time that we'll see more and more folks asking for static IP for that use case. The new `kubelet` API became [beta](https://github.com/kubernetes/enhancements/pull/4288) recently.

**Comment by mtl-wgtwo** [comment]
To chime in on this issue: we run telco workloads that require ingress and egress over fixed IPs, so this would be a very useful feature (and one that we have tested in Calico).

The primary constraint here is that we are using SCTP, which doesn't work well with load-balancers (traffic is loadbalanced using other means).

**Comment by xgp01** [comment]
I found this issue is implemented by @llhhbc in 2022, here is the PR https://github.com/cilium/cilium/pull/21379 .
For those who are interested in using this feature can refer to this pull request.

**Comment by MohammedNoureldin** [comment]
Hi @xgp01, I am a bit confused. Why did you close the issue if the feature is still officially not supported?

I mean you referred to a 2 years old PR, but for me this is not related to this.

If you were able to use a specific IP somehow, could you please provide more details?

**Comment by MohammedNoureldin** [comment]
Hi @gandro,

is this issue resolved by this pull request https://github.com/cilium/cilium/pull/34618?

**Comment by gcezaralmeida** [comment]
+1

**Comment by gandro** [comment]
> Hi [@gandro](https://github.com/gandro),
> 
> is this issue resolved by this pull request [#34618](https://github.com/cilium/cilium/pull/34618)?

Yes, at least in a bare-bones fashion: Thanks to that PR, it is now possible to create a pool in multi-pool mode which contains a single IP. 

I believe the UX could still be improved a bit (e.g. have Cilium automatically create the pool or similar and maybe think a bit about how to control allocations, i.e. ensuring that the static IPs are part of a preconfigured range or similar), so I think it makes sense to keep this open for now. But thanks to [#34618](https://github.com/cilium/cilium/pull/34618) it is possible to achieve static IPs in multi-pool mode by doing the following:

1. Create a new CiliumPodIPPool containing a singular IP, e.g. `ip-192-168-1-42`
2. Annotate your pod with `ipam.cilium.io/ip-pool: ip-192-168-1-42`

**Comment by liyihuang** [comment]
From static IP feature perspective, we should still keep the issue open since the current multi pool implementation only works with multi-pool not only for UX. If the use case is with on AWS with ENI mode or any other cloud native routing. this doesn't work that well.

**Comment by wanghonglei5181** [comment]
close

**Comment by davidhwua** [comment]
My use case is to run a static pod and static ip is very important for KubeVirt.

It wil be perfect f the cilium has the static ip function.

Is there any plan to develop the static ip function ?

**Comment by gandro** [comment]
Please read the above messages on how you can achieve static IPs already today with Multi-Pool IPAM. It's a bit clunky but it works.

Note however that specifically for KubeVirt live migration, static IP is not enough. I've written down the requirements here https://github.com/cilium/cilium/issues/38576#issuecomment-2788543187

I don't think there is anyone currently working (or planning to work) on adding this in the Cilium open source project.
